### PR TITLE
Add path config

### PR DESCRIPTION
**Pull request by MStokluska**
## Motivation
#294 

## Solution

Paths for `model` `schema` `generatedResolvers` and `customResolvers` can be configured in graphback.json with default values being the same as they were. 

## Verification

1. `graphback init <name>` to create graphback project
2. navigate to project and run `graphback generate`
3. graphback should work as usual
4. create new folder in you graphback project called `foo` and move models to `foo` folder
5. in `graphback.json` change model path to `/foo/model/` and `graphback generate` and `graphback db` should work as usual
6. in `graphback.json` change the output location for resolvers and run `graphback generate` resolvers should be generated into a new path
7. in `graphback.json` change the output of `client` and run `graphback generate`, `client` queries should be generated in a specified path
8. in `graphback.json` change output path of `schema`, `graphback generate` should create schema in a new folder
9. if any of new folders don't exist a folder should be created
10. if folder already exists it should not be overwritten but instead, files generated should be added to existing foler.

### CODE REVIEW

File: packages/graphback-cli/src/helpers/db.ts [hunk 1]
```diff
@@ -6,23 +6,23 @@ import { DatabaseSchemaManager, GraphQLBackendCreator } from 'graphback';
...
   }
   process.exit(0)
 }
 
-export const dropDBResources = async(): Promise<void> => {
+export const dropDBResources = async (): Promise<void> => {
   try {
     const { database, dbConfig } = JSON.parse(readFileSync(configPath, 'utf8'))
-    if(database === 'sqlite3') {
+    if (database === 'sqlite3') {
```
Comment (wtrocki) [inline packages/graphback-cli/src/helpers/db.ts]: Wrong formatting. It is strange to see this as we all using visual studio code.

File: packages/graphback-cli/src/helpers/common.ts [hunk 2]
```diff
@@ -1,12 +1,27 @@
...
+
+    const { paths } = JSON.parse(readFileSync(configPath, "utf8"))
+  
+    pathForModel = join(process.cwd(), paths.model)
+
+    accessSync(`${pathForModel}`)
+
+  } catch (err) {
+    logError(`Model not found, make sure you are in root directory of your project and that you have
+    specified correct path to your .graphql file`)
```
Comment (craicoverflow) [inline packages/graphback-cli/src/helpers/common.ts]:
```suggestion
    specified the correct path to your .graphql file`)
```

File: packages/graphback-cli/src/helpers/common.ts [hunk 3]
```diff
@@ -1,12 +1,27 @@
...
+    const configPath = `${process.cwd()}/graphback.json`
+
+    const { paths } = JSON.parse(readFileSync(configPath, "utf8"))
+  
+    pathForModel = join(process.cwd(), paths.model)
+
+    accessSync(`${pathForModel}`)
+
+  } catch (err) {
+    logError(`Model not found, make sure you are in root directory of your project and that you have
```
Comment (wtrocki) [inline packages/graphback-cli/src/helpers/common.ts]: Not issue with your code itself but it can be improved. 
Code above can fail in 2 places:

- on reading config - then we need to say we have a config issue
- on reading graphql files in the model

File: packages/graphback-cli/src/helpers/db.ts [hunk 4]
```diff
@@ -37,23 +37,23 @@ export const dropDBResources = async(): Promise<void> => {
   }
 }
 
-export const createDBResources = async(): Promise<void> => {
-  try{
-    const models = new GlobSync('model/*.graphql', { cwd: process.cwd()})
+export const createDBResources = async (pathForModel: string): Promise<void> => {
+  try {
+
```
Comment (wtrocki) [inline packages/graphback-cli/src/helpers/db.ts]: Let's avoid extra linebreaks etc.

File: packages/graphback-cli/src/helpers/db.ts [hunk 5]
```diff
@@ -75,10 +75,10 @@ Run ${chalk.cyan(`npm run develop`)} to start the server.
   `)
 }
 
-export const createDB = async(): Promise<void> => {
-  checkDirectory()
+export const createDB = async (): Promise<void> => {
+  const pathForModel = checkDirectory()
```
Comment (wtrocki) [inline packages/graphback-cli/src/helpers/db.ts]: It will be cool if we calculate all paths in single place and pass entire object instead of individual paths.
This can be separate PR though.

File: packages/graphback-cli/src/helpers/db.ts [hunk 6]
```diff
@@ -6,7 +6,7 @@ import { DatabaseSchemaManager, GraphQLBackendCreator } from 'graphback';
 import { logError, logInfo } from '../utils'
 import { checkDirectory } from './common'
 
-const configPath = `${process.cwd()}/config.json`
+const configPath = `${process.cwd()}/graphback.json`
```
Comment (wtrocki) [inline packages/graphback-cli/src/helpers/db.ts]: [Idea] Can we have function that will append all `process.cwd()` to all paths in config. Then we can have interface for it and pass it to every module to use it. This will be the only one `extra field` that we will need to require from outside as it is path to config itself.

File: packages/graphback-cli/src/templates/configTemplates.ts [hunk 7]
```diff
@@ -37,6 +37,14 @@ const generationConfig = {
   "disableGen": false
 }
 
+const paths = {
```
Comment (wtrocki) [inline packages/graphback-cli/src/templates/configTemplates.ts]:
```suggestion
const defaultPaths = {
```

File: packages/graphback-cli/src/helpers/common.ts [hunk 8]
```diff
@@ -1,12 +1,27 @@
...
-    accessSync(`${process.cwd()}/model`)
-  } catch(err) {
-    logError(`model directory not found. Make you sure you are in the root of your project.`)
+export const checkDirectory = (): string => {
+
+  let pathForModel
+
+  try {
+
+    const configPath = `${process.cwd()}/graphback.json`
```
Comment (wtrocki) [inline packages/graphback-cli/src/helpers/common.ts]: `${process.cwd()}/graphback.json` is used in couple places maybe we can setup it as constant in separate folder and simply require it?

File: templates/apollo-rest-starter/graphback.json [hunk 9]
```diff
@@ -17,5 +17,12 @@
     "subDelete": false,
     "disableGen": false
   },
+  "paths": {
+    "model": "./model",
+    "generatedResolvers": "./src/resolvers",
+    "customResolvers": "./src/resolvers",
+    "schema": "./src/schema",
+    "client": "./client"
```
Comment (wtrocki) [inline templates/apollo-rest-starter/graphback.json]:
```suggestion
    "client": "./client/src/graphql"
```

File: templates/apollo-rest-starter/src/config/config.ts [hunk 10]
```diff
@@ -1,5 +1,5 @@
 import knex from 'knex'
-import * as config from '../../config.json'
+import * as config from '../../graphback.json'
```
Comment (wtrocki) [inline templates/apollo-rest-starter/src/config/config.ts]: We can create issue about this problem to see if we can decompose config from the database.

File: packages/graphback-cli/src/helpers/generate.ts [hunk 11]
```diff
@@ -21,61 +22,65 @@ followed by ${chalk.cyan(`${cliName}db`)} to create database.
  * Generate schema and resolvers using graphback-core and
  * write them into generated folder
  */
-export async function generateBackend(): Promise<void> {
+export async function generateBackend(pathForModel: string): Promise<void> {
```
[the remark this answers is not in the document]
Comment (wtrocki) [reply packages/graphback-cli/src/helpers/generate.ts]: See comment above

**Comment by wtrocki** [comment]
> It is possible to set the paths outside of the project. For example:

We are making sure that we are always using CWD. People can use `../` which do not make sense as your source code or model will not be included into the project, but if they want it they could do it. There is no sense to invest time preventing some 0.000001% use case that someone might actually found useful :D

**Comment by wtrocki** [comment]
> It might be nice to be able to define what paths you want on graphback init. So something like:

Nice idea! @MStokluska this could be separate PR/Issue depending on your judgment.

File: packages/graphback-cli/src/helpers/generate.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (MStokluska) [reply packages/graphback-cli/src/helpers/generate.ts]: 100% going to investigate

File: packages/graphback-cli/src/templates/configTemplates.ts [hunk 12]
```diff
@@ -37,6 +37,14 @@ const generationConfig = {
...
+  "model": "./model",
+  "generatedResolvers": "./src/resolvers",
+  "customResolvers": "./src/resolvers",
+  "schema": "./src/schema",
+  "client": "./client/src/graphql"
```
Comment (ankitjena) [inline packages/graphback-cli/src/templates/configTemplates.ts]: This will throw error since `mkdir` creates folder if parent folder exists. To be precise the error
```
Error: ENOENT: no such file or directory, mkdir '/home/ankitj/apollo/client/src/graphql'
```
Are we assuming the user already has the folder structure created?

Comment (wtrocki) [reply packages/graphback-cli/src/templates/configTemplates.ts]: I think we can create a structure if it is not there.

Comment (MStokluska) [reply packages/graphback-cli/src/templates/configTemplates.ts]: hi there,
Ankit did you check this? We are first mkdir with recursive { true } which solves the issue.

Comment (ankitjena) [reply packages/graphback-cli/src/templates/configTemplates.ts]: @MStokluska I checked a bit and seems to be an issue the node versioning. I had downgraded my node version to v9.1.1 for some reason. The recursive true option works for anything above v10

File: packages/graphback-cli/src/helpers/generate.ts [hunk 13]
```diff
@@ -21,61 +22,65 @@ followed by ${chalk.cyan(`${cliName}db`)} to create database.
...
       }
     }
 
-    writeFileSync(outputSchemaPath, generated.schema)
-
-    writeFileSync(`${outputResolverPath}/index.ts`, generated.resolvers.index)
-    
-    if(!existsSync(`${outputResolverPath}/custom`)) {
-      mkdirSync(`${outputResolverPath}/custom`)
+    if(!existsSync(`${pathForSchema}`)) {
```
Comment (ankitjena) [inline packages/graphback-cli/src/helpers/generate.ts]:
```suggestion
    if(!existsSync(pathForSchema)) {
```

File: packages/graphback-cli/src/helpers/generate.ts [hunk 14]
```diff
@@ -21,61 +22,65 @@ followed by ${chalk.cyan(`${cliName}db`)} to create database.
...
+      mkdirSync(`${pathForSchema}`, { recursive: true })
```
Comment (ankitjena) [inline packages/graphback-cli/src/helpers/generate.ts]:
```suggestion
      mkdirSync(pathForSchema, { recursive: true })
```

**Comment by ankitjena** [comment]
@MStokluska Works perfect. Good work on this

File: packages/graphback-cli/src/helpers/generate.ts [hunk 15]
```diff
@@ -21,61 +22,65 @@ followed by ${chalk.cyan(`${cliName}db`)} to create database.
...
     }
-
-    if(!existsSync(`${outputResolverPath}/generated`)) {
-      mkdirSync(`${outputResolverPath}/generated`)
+    if(!existsSync(`${customResolvers}/custom`)) {
```
Comment (ankitjena) [inline packages/graphback-cli/src/helpers/generate.ts]: Why not use `./src/resolvers/custom` directly in paths instead of appending it here. Same for generated resolvers.

File: templates/graphql-js-starter/graphback.json [hunk 16]
```diff
@@ -17,5 +17,12 @@
     "subDelete": false,
     "disableGen": false
   },
+  "defaultPaths": {
```
Comment (darahayes) [inline templates/graphql-js-starter/graphback.json]: Minor question/suggestion. Is `defaultPaths` the name that will be present in the `graphback.json` file that users will modify? As an end user, the name `defaultPaths` is kinda weird because they're not default paths right? They're paths that I can change. Maybe something like `paths` or `files` would make a little more sense.

**Comment by wtrocki** [comment]
@ankitjena @MStokluska do you mind takling a look. I'm not sure about config format. 
Still need to update some docs.

**Comment by MStokluska** [comment]
> @darahayes @ankitjena @MStokluska do you mind takling a look. I'm not sure about config format.
> Still need to update some docs.

I like the new config a lot. I can see I've been on a good track with interfaces myself, just noticed thanks to your PR what I have been doing wrong! Looks like a really great job! Thanks @wtrocki

File: packages/graphback-cli/src/helpers/generate.ts [hunk 17]
```diff
@@ -23,73 +25,68 @@ followed by ${chalk.cyan(`${cliName}db`)} to create database.
...
         generatedClient[folder].forEach((c: ClientImplementation) => writeFileSync(`${currentFolder}/${c.name}.ts`, c.implementation))
       })
     }
-
   } catch (err) {
     logError(err)
     process.exit(0)
   }
+
+  function checkAndCreateFolders(pathForSchema: string, customResolvers: string, generatedResolvers: string) {
```
Comment (craicoverflow) [inline packages/graphback-cli/src/helpers/generate.ts]: This could be made more generic:

```js
function checkAndCreateFolders(...paths: string[]) {
	for const path in paths {
       if (!existsSync(path)) {
          mkdirSync(path, { recursive: true });
       }
	}
}
```

Comment (wtrocki) [reply packages/graphback-cli/src/helpers/generate.ts]: Good idea!

Comment (darahayes) [reply packages/graphback-cli/src/helpers/generate.ts]: Could throw an error mid loop if there were permissions problems

Comment (wtrocki) [reply packages/graphback-cli/src/helpers/generate.ts]: So IMHO let's not change that but.. also try catch it. 
@darahayes is an walking node.js reference :)

**Comment by wtrocki** [comment]
#worksforme

File: packages/graphback-cli/src/helpers/generate.ts [hunk 18]
```diff
@@ -23,73 +25,68 @@ followed by ${chalk.cyan(`${cliName}db`)} to create database.
...
     generated.resolvers.custom.forEach((output: OutputResolver) => {
-      if(!existsSync(`${outputResolverPath}/custom/${output.name}.ts`) || output.name === 'index') {
-        writeFileSync(`${outputResolverPath}/custom/${output.name}.ts`, output.output)
+      if (!existsSync(`${customResolvers}/${output.name}.ts`) || output.name === 'index') {
+        writeFileSync(`${customResolvers}/${output.name}.ts`, output.output)
       }
     })
 
-    const resolverFiles = readdirSync(`${outputResolverPath}/generated`)
-    resolverFiles.forEach((file: string) => unlinkSync(`${outputResolverPath}/generated/${file}`))
```
Comment (ankitjena) [inline packages/graphback-cli/src/helpers/generate.ts]: This was for the case that the user changes his types. So there would be no need for a resolver file from the older model.

File: packages/graphback-cli/src/config/ProjectConfig.ts [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+    /**
+     * Configuration for input/output folders
+     */
+    folders: FileConfig
+
+    /**
+     * Empty object config that can be created to generate client
+     * Reserved for future usage
+     */
+    client: object
```
Comment (ankitjena) [inline packages/graphback-cli/src/config/ProjectConfig.ts]: Shouldn't this be `boolean` instead of `object`

Comment (ankitjena) [inline packages/graphback-cli/src/config/ProjectConfig.ts]:
```suggestion
    client: boolean
```

File: packages/graphback-cli/src/helpers/generate.ts (hunk 18 above)
Comment (wtrocki) [reply packages/graphback-cli/src/helpers/generate.ts]: Removing people's code - that they could have made modifications on is risky business. Especially that changes might not be committed. 
So people will be able to manually remove them later and there is no need for them to exist due to the fact that they will not be required in index. It is tradeoff for now. 

More work will be done in #340

File: packages/graphback-cli/src/config/ProjectConfig.ts (hunk 19 above)
Comment (wtrocki) [reply packages/graphback-cli/src/config/ProjectConfig.ts]: One rule in building software that people depend on (libraries and frameworks) is to not introduce breaking changes at the beginning. I know that we assign boolean but if we will have any other field for the client later it will require people to change all their configs. 

So I have decided to use object as form of enablement and did not change underlying implementation from boolean.

**Comment by wtrocki** [comment]
So the only outstanding element here is to add extra try catch for fs/path methods.

**Comment by craicoverflow** [comment]
I cannot complete step 1. `graphback init` fails.

Screenshot from 2019-09-17 12-39-10

**Comment by MStokluska** [comment]
> I cannot complete step 1. `graphback init` fails.
> 
> Screenshot from 2019-09-17 12-39-10

Hey, I'm having the same issue.

File: packages/graphback-cli/src/config/ConfigBuilder.ts [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Abstraction for Graphback configuration
+ */
+export class ConfigBuilder {
+
+    public config: ProjectConfig
+
+    constructor(creator: () => ProjectConfig) {
+        this.config = creator()
+        this.applyAbsolutePaths();
```
Comment (ankitjena) [inline packages/graphback-cli/src/config/ConfigBuilder.ts]: We should not be calling this here in `init`, this can be done later in `generate`. Moreover `process.cwd()` is different during `init`. This causes init to fail

**Comment by wtrocki** [comment]
Have all fixed. Just need more time for tests
